jubilant/_juju.py

```diff
@@ -117,28 +118,57 @@ def status(
         result = json.loads(stdout)
         return Status.from_dict(result)
 
-    def wait_status(
+    def wait(
         self,
-        ready_func: typing.Callable[[Status], bool],
+        ready: typing.Callable[[Status], bool],
         *,
         model: str | None = None,
-        timeout: float = 10 * 60,
-        delay: float = 1,
+        # TODO: should this callable raise instead? maybe better errors
+        error: typing.Callable[[Status], bool] = any_error,
+        delay: float = 1.0,
+        timeout: float = 10 * 60.0,
         successes: int = 3,
     ) -> Status:
-        """TODO."""
-        start = time.time()
+        """Wait until ``ready(status)`` returns true, *successes* times in a row.
+
+        This fetches the Juju status repeatedly (waiting *delay* seconds between each call),
+        and returns the last status after the *ready* callable returns true a number
+        of times in a row (*successes* times).
+
+        Args:
+            ready: Callable that takes a :class:`Status` object and returns true when the wait
+                should be considered ready. It needs to return true *successes* times in a row
+                before ``wait`` returns.
+            model: Juju model name. Overrides ``self.model`` if that is set.
```

Comment:
It's an interesting question. You could eliminate the `model` argument on most of the methods this way. One reason to keep them would be for `juju = Juju()` "pure" command line usage. Kinda contrived, but for example:

```python
juju = Juju()  # no model specified, "pure" 1:1 to juju command

juju.add_model('t1')
juju.add_model('t2')

juju.deploy('charm1', model='t1')
juju.deploy('charm2', 'model='t2')

juju.wait(lambda s: s.is_app_active('charm1'), model='t1')
juju.wait(lambda s: s.is_app_active('charm2'), model='t2')
```

Reply:
Yeh I don't like this. 
I'd much rather do 

```
cos_juju = Juju(model="cos")
openstack_juju = Juju(model="openstack")

cos_juju.,deploy("traefik")
openstack_juju.deploy(...)
...
```

Reply:
Both of these feel like "pure" command line usage to me. Ben's is like CLI usage where you're putting `-m` in a lot. Pietro's is like CLI usage where you `switch` (but instead of calling `switch()` you switch the name you call from).

I'm obviously not 'real' Juju admin, but I generally switch, which is why I lean towards the style Pietro outlined, I think. I'm curious if we know what people who are actually running controllers in production tend to do.

Reply:
That makes sense to me, and is simpler. Can always add the `model` arg overrides later if needed.

Reply:
I've updated this so there are no `model` override args to `deploy`, `status`, etc. However, they're still required for `add_model`, `destroy_model` (and `switch`). I'd prefer the type signatures to be simpler and there not be too many ways of doing things:

```python
# To use the current model
juju = jubilant.Juju()  # self.model = None
juju.deploy('mycharm')

# To create a new model (probably most common for integration tests)
juju = jubilant.Juju()
juju.add_model('newmodel')  # also sets self.model
juju.deploy('mycharm')  # executed with "--model mymodel"

# To use an existing model
juju = jubilant.Juju(model='existingmodel')  # sets self.model
juju.deploy('mycharm')
```